File: week0003/pertsodian/java/src/fintech/domain/Account.java
package fintech.domain;

import java.math.BigDecimal;

import fintech.exceptions.InsufficientBalanceException;

public class Account {

  private final String name;
  private BigDecimal balanceAmount;
  
  public Account(String name, String initialAmount) {
    if (name == null)
      throw new IllegalArgumentException("Invalid input name");
    this.name = name;
    this.balanceAmount = parseAmount(initialAmount);
  }
  
  public String getName() {
    return name;
  }
  
  public BigDecimal getCurrentBalance() {
    return balanceAmount;
  }
  
  public void deposit(String amount) {
    adjustBalance(parseAmount(amount));
  }
  
  public void withdraw(String amount) {
    adjustBalance(parseAmount(amount).negate());
  }
  
  private void adjustBalance(BigDecimal amount) {
    BigDecimal adjustedBalance = balanceAmount.add(amount);
    if (adjustedBalance.compareTo(BigDecimal.ZERO) < 0)
      throw new InsufficientBalanceException();
    balanceAmount = adjustedBalance; 
  }

  private BigDecimal parseAmount(String amountStr) {
    try {
      BigDecimal amount = new BigDecimal(amountStr);
      if (isNegative(amount))
        throw new IllegalArgumentException("Amount must be positive");
      return amount;
    }
    catch (NumberFormatException e) {
      throw new IllegalArgumentException("Invalid input amount");
    }
  }
  
  private boolean isNegative(BigDecimal amount) {
    return amount.compareTo(BigDecimal.ZERO) <= 0;
  }
}

File: week0003/mattfanz/php/src/FinTech/Account.php
<?php
/*
 * This file is part of the FinTech.
 */
namespace FinTech;

/**
 * Class Account
 * @package FinTech
 */
class Account
{
    /**
     * @var string
     */
    protected $name;

    /**
     * @var double
     */
    protected $amount = 0.00;

    /**
     * Account constructor.
     * @param $name
     * @param $amount
     */
    public function __construct($name, $amount)
    {
        $this->name = $name;
        $this->amount = (double) $amount;
    }

    /**
     * @param $amount
     * @return bool
     */
    public function deposit($amount)
    {
        if ($this->checkTransaction($amount)) {
            $this->amount += $amount;
            return true;
        }
        return false;
    }

    /**
     * @param $amount
     * @return bool
     */
    public function withdraw($amount)
    {
        if ($this->checkTransaction($amount) && $this->amount - $amount >= 0) {
            $this->amount -= $amount;
            return true;
        }
        return false;
    }

    /**
     * @param $amount
     * @return bool
     */
    protected function checkTransaction($amount)
    {
        return (is_numeric($amount) && $amount > 0);
    }
}
